### DISCUSSION TITLE
Restricting string length

### DISCUSSION THREAD

**Discussion by khepin** [cuelang/cue#657] [Q&A] [answered]
How would I put a constraint on a string's length? Right now I've only figured out how to do it via regexes:

```cue
#StrLen8: =~"^.{8}$"
{
    v: #StrLen8
}
```
```json
{
    "v": "hello123"
}
```
This also requires 1 regex / length unless I'm missing something...

**Comment by mpvl** [comment, marked as answer]
```
import "strings"

v: strings.MaxRunes(8) & strings.MinRunes(8)
```
There is a PR pending to add `NumRunes` too.

But you could make regex's variable:
```
#StrLen: {
    =~"^.{\(#n)}$"
   #n: int
}
#StrLen8: { #StrLen, #n: 8 }
```
Scalar values may have definition fields associated with them. This is a fairly new feature and some syntax has been proposed to make this more natural.

**Comment by mpvl** [comment]
Note that the `let` builtin deliberately does not work on strings. String length is kind of a complicated matter, so we decided that one needs to be very explicit about the kind of string length that one is using.

**Reply by khepin** [reply]
Good point on string length not being straightforward. Easy to overlook. So from previous answers, I imagine `MaxRunes` counts UTF8 runes. And the regex version here would count `bytes` right?

**Reply by extemporalgenome** [reply]
> I imagine MaxRunes counts UTF8 runes

There shouldn't be any distinction between "unicode runes" and "utf-8 runes" since _rune_ is a semantic concept, regardless of the underlying encoding.

> String length is kind of a complicated matter, so we decided that one needs to be very explicit about the kind of string length that one is using.

Could we just have a default meaning for `bytes` vs `string`, such that `len('...')` always returns the byte-length, while `len("...")` returns the number of runes, leaving the very-explicit functions for the job of counting bytes in a `string` or counting runes in a `bytes`?

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/657.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
